kafka_consumer/check.py

```diff
@@ -29,100 +40,266 @@ def __init__(self, name, init_config, agentConfig, instances=None):
         self.kafka_timeout = int(
             init_config.get('kafka_timeout', DEFAULT_KAFKA_TIMEOUT))
 
-    def check(self, instance):
-        consumer_groups = self.read_config(instance, 'consumer_groups',
-                                           cast=self._validate_consumer_groups)
-        zk_connect_str = self.read_config(instance, 'zk_connect_str')
-        kafka_host_ports = self.read_config(instance, 'kafka_connect_str')
+    def _get_highwater_offsets(self, kafka_hosts_ports):
+        """
+        Fetch highwater offsets for each topic/partition from Kafka cluster.
+
+        Do this for all partitions in the cluster because even if it has no
+        consumers, we may want to measure whether producers are successfully
+        producing. No need to limit this for performance because fetching broker
+        offsets from Kafka is a relatively inexpensive operation.
+
+        Sends one OffsetRequest per broker to get offsets for all partitions
+        where that broker is the leader:
+        https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-OffsetAPI(AKAListOffset)
+        """
+        # We could switch this to a long-lived connection that's cached
+        # on the AgentCheck instance, in which case we don't have to
+        # cleanup the connection. Downside is there's no way to cleanup
+        # connection when agent restarts, so brokers before 0.9 just
+        # accumulate stale connections. In 0.9 Kafka added
+        # connections.max.idle.ms https://issues.apache.org/jira/browse/KAFKA-1282
+        kafka_client = KafkaClient(
+            bootstrap_servers=kafka_hosts_ports,
+            request_timeout_ms=(self.kafka_timeout * 1000))
+        try:
+            highwater_offsets = {}
+            # store partitions that exist but unable to fetch offsets for later
+            # error checking
+            topic_partitions_without_a_leader = []  # contains (topic, partition) tuples
+            for broker in kafka_client.cluster.brokers():
+                if not kafka_client.ready(broker.nodeId):
+                    time.sleep(0.5)
```

Comment:
This is a major hack, we had a discussion about better ways to do this (and the problems with them) here: https://github.com/DataDog/integrations-core/pull/271#discussion_r118064359